design/one-pager-generic-usage-type.md

```diff
@@ -0,0 +1,410 @@
+be achieved by adding an owner reference from the using resource to the `Usage`.
+
+**Preventing early deletion of the `Usage`:**
+
+Typically, we expect a `Usage` resource to be defined as part of the same
+composition as the using and used resources. This means the `Usage` will get
+the delete request when the composite resource is deleted. In this case, the
+controller will prevent the deletion of the `Usage` until the using resource is
+deleted with the help of a finalizer.
+
+**Defining admission webhook rules:**
+
+The admission webhook will intercept the `DELETE` requests for the used
+resource(s). To achieve this, the controller will add [webhook rules] for the
+used resource(s) when a `Usage` is created. It is also the controller's job to
+remove these rules when the `Usage` is deleted, and no other `Usage` resources
+reference the same used resource(s).
+
+#### Admission Webhook
+
+The admission webhook will be responsible for intercepting the `DELETE` requests
+for the used resource(s) and rejecting them with a ["409 Conflict" error code].
+We will fail open in case of any errors from the webhook with
+[`failurePolicy: Fail`].
+
+The webhook will be implemented as a Kubernetes validating admission webhook
+leveraging the [existing webhook machinery] in Crossplane and will be running
+as part of the Crossplane core.
+
+We will use the [Field Indexers] to efficiently query the used resource(s) when
+a `DELETE` request is received.
+
+Implementation of the core logic for the webhook handler:
+
+```go
+func (h *handler) validateNoUsages(ctx context.Context, u *unstructured.Unstructured) admission.Response {
+	usageList := &v1alpha1.UsageList{}
+	if err := h.reader.List(ctx, usageList, client.MatchingFields{inUseIndexKey: getIndexValueForObject(u)}); err != nil {
+		return admission.Errored(http.StatusInternalServerError, err)
+	}
+	if len(usageList.Items) > 0 {
+		return admission.Response{
+			AdmissionResponse: admissionv1.AdmissionResponse{
+				Allowed: false,
+				Result: &metav1.Status{
+					Code:   int32(http.StatusConflict),
+					Reason: metav1.StatusReason(fmt.Sprintf("The resource is used by %d resource(s), including %s/%s", len(usageList.Items), usageList.Items[0].Spec.By.Kind, usageList.Items[0].Spec.By.Name)),
+				},
+			},
+		}
+	}
+	return admission.Allowed("")
+}
+```
+
+### User Experience
+
+#### Defining a Usage Resource as part of a Composition
+
+We expect a typical usage relationship to be defined as part of the same
+composition as the using and used resources. In this case, the user will
+define the `Usage` resource as part of the composition template.
+
+Let's take [upbound/platform-ref-aws] as an example. In this composition, we
+have a [composition for `XCluster`] which composes `XEKS` composing an EKS cluster,
+and `XServices` composing a set of helm releases running on the cluster. In this
+case, we will add the following resource to the same composition to define a
```

Comment:
disclaimer: i haven't thought through this problem space really at all, so I'm not saying this is possible at all 😊 

i was naively hoping that these usage relationships could get automatically determined by the crossplane machinery - i.e. there would be no additional burden on the users to figure out, declare, and manage these `Usage` objects themselves.

Did we already write down why that simplification of the user experience is not possible? would be curious to read more about that 😇

Reply:
Crossplane has a very narrow view of the dependency, limited to Crossplane-native and kubernetes-native objects like `ProviderConfigs` and `Secrets`.  Dependencies on Composites are not really visible to Crossplane.  For example Crossplane "knows" that a `Release` object has a `ProviderConfig` which depends on a `Secret` that contains the cluster credentials, but it has no idea that the helm chart modeled by the `Release` also has implicit dependencies on the Cluster, Nodegroup and all of the other resources that make up the "Cluster Composite".  The logical dependency is actually on everything that comprises a running cluster, not just on the credentials in the Secret.  The author of the composition is really the only one who "knows" what the actual dependency is.

That being said, it would definitely be helpful to the composition author to have a shortcut notation for the `Usage` that does not involve defining the entire resource as an entry in the resource list.  Something along the lines of a terraform `dependsOn` attribute that could cause Crossplane to automatically generate the `Usage` object, but I understand the reluctance to extend the DSL to include more terraform-like syntax.  Another possibility is to use a Composition Function to generate the `Usage` objects based on one or more annotations defined in the resource or passed in using the Composition Functions `config` object.

I think this could be part of a library of common Composition Functions that provide specific functionality on top of the standard DSL.